File: client/python/example/quick/testUser.py
import grpc
import user_pb2
import finger_pb2

from example.user.user import UserSvc

NUM_OF_NEW_USER = 3
START_USER_ID = 10000000
TEMPLATE_FORMAT = finger_pb2.TEMPLATE_FORMAT_SUPREMA
QUALITY_THRESHOLD = 50

def getNewUserIDs():
  userIDs = []

  for i in range(0, NUM_OF_NEW_USER):
    userIDs.append(str(START_USER_ID + i))

  return userIDs


def makeNewUser():
  userInfos = []

  for i in range(0, NUM_OF_NEW_USER):
    userHdr = user_pb2.UserHdr(ID=str(START_USER_ID + i))
    userInfo = user_pb2.UserInfo(hdr=userHdr)

    userInfos.append(userInfo)

  return userInfos

def testSetFinger(userSvc, fingerSvc, deviceID, userID):
  userIDs = [userID]

  try:
    users = userSvc.getUser(deviceID, userIDs)

    print(f'\nUser without fingerprint: \n{users[0]}')

    print('>>> Scan a finger...', flush=True)

    templateData1 = fingerSvc.scan(deviceID, TEMPLATE_FORMAT, QUALITY_THRESHOLD)

    print('>>> Scan the same finger again...', flush=True)

    templateData2 = fingerSvc.scan(deviceID, TEMPLATE_FORMAT, QUALITY_THRESHOLD)

    fingerData = finger_pb2.FingerData(templates=[templateData1, templateData2])
    userFingers = [user_pb2.UserFinger(userID=userID, fingers=[fingerData])]

    userSvc.setFinger(deviceID, userFingers)

    users = userSvc.getUser(deviceID, userIDs)

    print(f'\nUser with fingerprint: \n{users[0]}')

  except grpc.RpcError as e:
    print(f'Cannot complete the set finger test: {e}')    


def testUser(userSvc, fingerSvc, deviceID):
  try:
    userList = userSvc.getList(deviceID)
    print(f'Initial user list: {userList}', flush=True)

    userIDs = []
    for user in userList:
      userIDs.append(user.ID)

    userInfos = userSvc.getUser(deviceID, userIDs)

    for info in userInfos:
      print(f'\nUser: \n{info}', flush=True)

    userSvc.enroll(deviceID, makeNewUser(), True)

    userList = userSvc.getList(deviceID)
    print(f'User list after adding new users: {userList}', flush=True)

    testSetFinger(userSvc, fingerSvc, deviceID, getNewUserIDs()[0])

    userSvc.delete(deviceID, getNewUserIDs())

    userList = userSvc.getList(deviceID)
    print(f'User list after deleting new users: {userList}', flush=True)

  except grpc.RpcError as e:
    print(f'Cannot complete the user test: {e}')    

File: client/node/example/quick/testUser.js
const user = require('../user');
const finger = require('../finger');

const NUM_OF_NEW_USER = 3;
const START_USER_ID = 10000000;
const TEMPLATE_FORMAT = finger.TEMPLATE_FORMAT_SUPREMA;

function getNewUserIDs() {
  var userIDs = [];

  for(i = 0; i < NUM_OF_NEW_USER; i++) {
    userIDs.push(`${START_USER_ID + i}`);
  }

  return userIDs;
}


function makeNewUser() {
  var userInfos = [];

  for(i = 0; i < NUM_OF_NEW_USER; i++) {
    let userInfo = new user.userMessage.UserInfo();
    let userHdr = new user.userMessage.UserHdr();
    userHdr.setId(`${START_USER_ID + i}`);
    userInfo.setHdr(userHdr);

    userInfos.push(userInfo);
  }

  return userInfos;
}

function testSetFinger(devID, userID) {
  var fingerData = new finger.fingerMessage.FingerData();

  var userIDs = []
  userIDs.push(userID);
  
  return user.getUser(devID, userIDs)
  .then((userInfos) => {
    console.log('User without fingerprint: ', userInfos[0]);

    console.log('>>> Scan a finger');

    return finger.scan(devID, TEMPLATE_FORMAT, 50);
  })
  .then((templateData) => {
    fingerData.addTemplates(templateData, 0);

    console.log('>>> Scan the same finger again');

    return finger.scan(devID, TEMPLATE_FORMAT, 50);
  })
  .then((templateData) => {
    fingerData.addTemplates(templateData, 1);

    var userFingers = [];
    var userFinger = new user.userMessage.UserFinger();
    userFinger.setUserid(userID);
    userFinger.addFingers(fingerData, 0);
    userFingers.push(userFinger);

    return user.setFinger(devID, userFingers);
  })
  .then(() => {
    return user.getUser(devID, userIDs);
  })
  .then((userInfos) => {
    console.log('User with fingerprint: ', userInfos[0]);
  })
  .catch((err) => {
    console.error('Cannot finish the setFinger test: ', err);
    throw err;
  });
}


function test(devID) {
  return user.getList(devID)
  .then((hdrs) => {
    console.log('User list: ', hdrs);

    var userIDs = [];

    for(i = 0; i < hdrs.length; i++) {
      userIDs.push(hdrs[i].id);
    }
    
    return user.getUser(devID, userIDs);
  })
  .then((userInfos) => {
    for(i = 0; i < userInfos.length; i++) {
      console.log('User: ', userInfos[i]);
    }

    return user.enroll(devID, makeNewUser());
  })
  .then(() => {
    return user.getList(devID);
  })
  .then((hdrs) => {
    console.log('User list after enrolling new users: ', hdrs);

    return testSetFinger(devID, getNewUserIDs()[0]);
  })
  .then(() => { 
    return user.deleteUser(devID, getNewUserIDs());
  })
  .then(() => {
    return user.getList(devID);
  })
  .then((hdrs) => {
    console.log('User list after deleting new users: ', hdrs);
  })
  .catch((err) => {
    console.error('Cannot finish the user test: ', err);
    throw err;
  });
}

module.exports.test = test;